cpp/src/table/row_operators.cu

```diff
@@ -373,51 +385,391 @@ namespace row {
 
 namespace lexicographic {
 
-std::shared_ptr<preprocessed_table> preprocessed_table::create(
-  table_view const& t,
+namespace {
+
+/**
+ * @brief Transform any nested lists-of-structs column into lists-of-integers column.
+ *
+ * For a lists-of-structs column at any nested level, its child structs column will be replaced by a
+ * `size_type` column computed as its ranks.
+ *
+ * If the input column is not lists-of-structs, or does not contain lists-of-structs at any nested
+ * level, the input will be passed through without any changes.
+ *
+ * @param lhs The input lhs column to transform
+ * @param rhs The input rhs column to transform (if available)
+ * @param column_null_order The flag indicating how nulls compare to non-null values
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ * @return A tuple consisting of new column_view representing the transformed input, along with
+ *         their ranks column(s) (of `size_type` type) and possibly new list offsets generated
+ *         during the transformation process
+ */
+std::tuple<column_view,
+           std::optional<column_view>,
+           std::vector<std::unique_ptr<column>>,
+           std::vector<std::unique_ptr<column>>>
+transform_lists_of_structs(column_view const& lhs,
```

Comment:
I think we should consider adding an overload for single tables that skips returning the unused pieces. There is a lot of branching in this function right now for one vs. two tables. We might be able to improve the awkwardness at the call site by having two overloads for one and two tables (which would be non-optional), and pushing the implementation into a function that is structured like the current state here. Just wrapping this function in a clearer one/two table overload would be helpful, I think.

Reply:
I thought about that too. But then didn't implement it because we may have potential optimization later (https://github.com/rapidsai/cudf/issues/12932). I plan to work on it after this PR. With that implemented, we will get rid of this branch entirely: https://github.com/rapidsai/cudf/pull/12953/files/3e4b7b2e22a332cef7b4473a69c65333b192f210#diff-c143ae2ecfe63796e604e5854f4a47cf6b831ebe5d68ce7c19a0e1533fceddc7R508-R509

Reply:
Okay. I am fine with deferring further work on this for now.